examples/upgradeable-contract/upgradeable-flipper/upgradeable.rs

```diff
@@ -0,0 +1,191 @@
+        if ink_env::get_contract_storage::<T>(ptr.advance_by(0))
+            .expect("could not properly decode storage entry")
+            .is_none()
+        {
+            <Self as SpreadAllocate>::allocate_spread(ptr)
+        } else {
+            Upgradeable::new(<T as SpreadLayout>::pull_spread(ptr))
+        }
+    }
+
+    fn push_spread(&self, ptr: &mut KeyPtr) {
+        <T as SpreadLayout>::push_spread(&self.inner, ptr)
+    }
+
+    fn clear_spread(&self, ptr: &mut KeyPtr) {
+        <T as SpreadLayout>::clear_spread(&self.inner, ptr)
+    }
+}
+
+/// Below the boilerplate code to implement `PackedLayout`, `SpreadAllocate`, `PackedAllocate`.
+
+impl<T: PackedLayout> PackedLayout for Upgradeable<T, Initialized> {
+    fn pull_packed(&mut self, at: &Key) {
+        <T as PackedLayout>::pull_packed(&mut self.inner, at)
+    }
+
+    fn push_packed(&self, at: &Key) {
+        <T as PackedLayout>::push_packed(&self.inner, at)
+    }
+
+    fn clear_packed(&self, at: &Key) {
+        <T as PackedLayout>::clear_packed(&self.inner, at)
+    }
+}
+
+impl<T: PackedLayout + SpreadAllocate> PackedLayout for Upgradeable<T, NotInitialized> {
+    fn pull_packed(&mut self, at: &Key) {
+        <T as PackedLayout>::pull_packed(&mut self.inner, at)
+    }
+
+    fn push_packed(&self, at: &Key) {
+        <T as PackedLayout>::push_packed(&self.inner, at)
+    }
+
+    fn clear_packed(&self, at: &Key) {
+        <T as PackedLayout>::clear_packed(&self.inner, at)
+    }
+}
+
+impl<T: SpreadAllocate + PackedLayout> SpreadAllocate for Upgradeable<T, Initialized> {
+    fn allocate_spread(ptr: &mut KeyPtr) -> Self {
+        Upgradeable::new(<T as SpreadAllocate>::allocate_spread(ptr))
+    }
+}
+
+impl<T: SpreadAllocate + PackedLayout> SpreadAllocate for Upgradeable<T, NotInitialized> {
+    fn allocate_spread(ptr: &mut KeyPtr) -> Self {
+        Upgradeable::new(<T as SpreadAllocate>::allocate_spread(ptr))
+    }
+}
+
+impl<T: PackedAllocate> PackedAllocate for Upgradeable<T, Initialized> {
+    fn allocate_packed(&mut self, at: &Key) {
+        <T as PackedAllocate>::allocate_packed(&mut self.inner, at)
+    }
+}
+
+impl<T: PackedAllocate> PackedAllocate for Upgradeable<T, NotInitialized> {
+    fn allocate_packed(&mut self, at: &Key) {
+        <T as PackedAllocate>::allocate_packed(&mut self.inner, at)
+    }
+}
+
+impl<T: PackedLayout, State> core::ops::Deref for Upgradeable<T, State> {
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
+
+impl<T: PackedLayout, State> core::ops::DerefMut for Upgradeable<T, State> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.inner
+    }
+}
+
+impl<T: PackedLayout, State> AsRef<T> for Upgradeable<T, State> {
+    #[inline]
+    fn as_ref(&self) -> &T {
+        &self.inner
+    }
+}
+
+impl<T: PackedLayout, State> AsMut<T> for Upgradeable<T, State> {
+    #[inline]
+    fn as_mut(&mut self) -> &mut T {
+        &mut self.inner
+    }
+}
```

Comment:
I know these impls are nice to have, but it doesn't look like they're used anywhere. For the sake of reducing boilerplate in the example I'd remove them